libs/pdi-scanner/src/rusb_async/pool.rs

```diff
@@ -0,0 +1,269 @@
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_interrupt(&mut self, endpoint: u8, buf: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::interrupt(self.device.as_raw(), endpoint, buf);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_iso(&mut self, endpoint: u8, buf: Vec<u8>, iso_packets: i32) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::iso(self.device.as_raw(), endpoint, buf, iso_packets);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    pub fn poll(&mut self, timeout: Duration) -> Result<CompletedTransfer> {
+        let next = self.pending.front().ok_or(Error::NoTransfersPending)?;
+        if poll_completed(self.device.context(), timeout, next.completed_flag()) {
+            let mut transfer = self.pending.pop_front().unwrap();
+            let endpoint = transfer.transfer().endpoint;
+            let data = transfer.handle_completed()?;
+            Ok(CompletedTransfer::new(endpoint, data))
+        } else {
+            Err(Error::PollTimeout)
+        }
+    }
+
+    pub fn poll_endpoint(&mut self, endpoint: u8, timeout: Duration) -> Result<Vec<u8>> {
+        for (index, transfer) in self.pending.iter_mut().enumerate() {
+            if transfer.transfer().endpoint == endpoint {
+                if poll_completed(self.device.context(), timeout, transfer.completed_flag()) {
+                    let mut transfer = self.pending.remove(index).unwrap();
+                    return transfer.handle_completed();
+                }
+
+                return Err(Error::PollTimeout);
+            }
+        }
+
+        Err(Error::NoTransfersPending)
+    }
+
+    pub fn cancel_all_for_endpoint(&mut self, endpoint: u8) {
+        for transfer in self.pending.iter_mut() {
+            if transfer.transfer().endpoint == endpoint {
+                transfer.cancel();
+            }
+        }
+    }
+
+    pub fn cancel_all(&mut self) {
+        // Cancel in reverse order to avoid a race condition in which one
+        // transfer is cancelled but another submitted later makes its way onto
+        // the bus.
+        for transfer in self.pending.iter_mut().rev() {
+            transfer.cancel();
+        }
+    }
+
+    /// Returns the number of async transfers pending
+    pub fn pending(&self) -> usize {
+        self.pending.len()
+    }
+}
+
+unsafe impl<C: UsbContext> Send for TransferPool<C> {}
+// unsafe impl<C: UsbContext> Sync for TransferPool<C> {}
```

Comment:
This wasn't commented out in the original, but I think I ran into problems with it so I removed it? I don't really remember.